File: samples/helloworld.js
/** A very basic proof-of-concept extension that resolves images to a directory defined in the config.
 *
 * The identifiers supported by this extension are filenames without an extension, but with a `js-` prefix.
 * They will be resolved to an image file without the prefix in the directory, if it exists.
 *
 * It must be configured with a `baseDirectory` in the extension configuration.
 */

// We provide a subset of the synchronous node.js 'fs' module here for convenience.
import { readDirSync } from "wolpi:fs";

const IMAGE_EXTENSIONS = new Set(['jpg', 'jpeg', 'png', 'gif', 'jp2', 'tif', 'webp']);

export default {
  // Define the extension metadata.
  info: () => ({
    apiVersion: 1,
    name: 'hello-world',
    description: 'just a simple resolving proof-of-concept'
  }),
  // The resolve function is called when an identifier needs to be resolved to an image source
  resolve: (identifier) => {
    if (!identifier.startsWith('js-')) {
      return;
    }
    identifier = identifier.substring(3); // Remove the 'js-' prefix.
    // The `wolpi` global provides access to the Wolpi context, including the configuration for the extensions.
    for (const { parentPath, name } of readDirSync(wolpi.config().baseDirectory)) {
      const basename = name.substring(0, name.lastIndexOf('.'));
      const extension = name.substring(name.lastIndexOf('.') + 1);
      if (basename === identifier && IMAGE_EXTENSIONS.has(extension)) {
        return { path: `${parentPath}/${name}` };
      }
    }
  }
}

File: samples/helloworld.py
"""A simple Wolpi extension that resolves image files by their `py-` prefixed filename in a directory.

The identifier (without prefix) is matched against the filename (without extension) in a specified base
directory. If a match is found for one of the supported image file extensions, the absolute path to the file
is returned.

"""
from pathlib import Path

from wolpi import ExtensionInfo, FilesystemImageSource, config

IMAGE_EXTENSIONS = {'.jpg', '.jpeg', '.png', '.gif', '.jp2', '.tif', '.webp'}

def info() -> ExtensionInfo:
    return {
        'apiVersion': 1,
        'name': 'hello-world-py',
        'description': 'just a simple resolving proof-of-concept'
    }

def cleanup():
    # No cleanup necessary for this simple extension, since we don't keep state
    # between hook invocations
    pass

def resolve(identifier: str, etag: str | None = None, last_modified: str | None = None) -> FilesystemImageSource | None:
    if not identifier.startswith('py-'):
        return
    identifier = identifier[3:]  # Remove 'py-' prefix
    # The `wolpi` module provides access to the Wolpi context, including the configuration for the extensions.
    base_dir = Path(config['baseDirectory'])
    for path in base_dir.iterdir():
        if path.stem == identifier and path.suffix in IMAGE_EXTENSIONS:
            return {'path': str(path.absolute())}
